xonsh/environ.py

```diff
@@ -7,62 +7,188 @@
 import builtins
 import platform
 import subprocess
+from datetime import datetime
+from functools import partial
 from warnings import warn
 
 from xonsh.tools import TERM_COLORS
 
-def current_branch(cwd=None):
-    """Gets the branch for a current working directory. Returns None
-    if the cwd is not a repository.  This currently only works for git, 
-    bust should be extended in the future.
-    """
-    branch = None
-    cwd = os.getcwd() if cwd is None else cwd
+class PromptFormatter(dict):
+    '''
+    Base class that implements utilities for PromptFormatters.
 
-    # step out completely if git is not installed
-    try:
-        binary_location = subprocess.check_output(['which', 'git'], cwd=cwd,
-                                    stderr=subprocess.PIPE,
-                                    universal_newlines=True)
-        if not binary_location:
-            return branch
-    except subprocess.CalledProcessError:
-        return branch
+    You should read the docs for this class to learn how to write your own
+    Prompt Formatter.  However, you most likely want to inherit from
+    DefaultPromptFormatter if you are writing a custom formatter.
+
+    Custom Prompt Formatters will need to implement their own __init__()
+    method that adds entries to several dicts depending on the needs of the
+    variables they add.  The Formatter may also implement methods to implement
+    those variables (although very simple variables can be written with
+    pre-existing functions.)
+
+    Example:
+
+    def __init__(self, *args, **kwargs):
+        super(PromptFormatter, self).__init__(*args, **kwargs)
+
+        # Add a variable that's just a static string.  Useful for constants
+        # that you can't remember:
+        self['bell'] = '\x07'
+
+        # To add a variable that only needs to be computed once add
+        # the function to be called to self.  This can be used for any
+        # variable that won't change while this shell instance is run.  It is
+        # especially useful if the variable takes a long time to compute and
+        # is infrequently used by users:
+        self['login_time'] = self.login_time
+
+        # To add a variable that needs to be computed every time the prompt is
+        # printed simply add a function to self and then add the variable name
+        # to the self._run_every set.
+        self['time24'] = functools.partial(time.strftime, '%H:%M:%S')
+        self._run_every.add('time24')
+
+    def login_time(self):
+        who_out = subprocess.check_output(['who', '-m'])
+        return ' '.join(who_out.decode().split()[2:4])
+    '''
+
+    def __init__(self, *args, **kwargs):
+
+        super(PromptFormatter, self).__init__(*args, **kwargs)
+        self._run_every = set()
```

Comment:
```_run-every`` needs to be marked with a single underscore.  In C++ terms, it can be considered a protected attribute.  It's part of the API if you are subclassing but it is not part of the API if you are calling from code outside of the inheritance tree.

Long winded explanation:  You need to have attributes of your class that are public, private, and protected in an OOP design.  python, gives us conventions to achieve privacy but the interpreter (mostly) doesn't enforce it.  So the conventions are guidance to other programmers on what their code should not touch in order to be a good citizen.  The convention for public attributes is that they should not start with an underscore.  Any programmer working on code unrelated to the class knows they should not touch attributes that start with an underscore but may touch any attributes that do not start with an underscore.  A programmer working on a class knows that they can touch the attributes of the class that begin with a leading underscore as long as they do not let too much of the implementation of that attribute leak into the outside world.  Now the question remains, what do python programmers need to do to implement protected attributes?  If you omit the leading underscore, programmers working on unrelated code will think that the attribute is public and also use it.  This is compounded by the fact that these programmers know the least about the code and thus are least able to make an informed decision about whether there's an alternative to that attribute, what the consequences are to using that attribute, and how to code a fix should that attribute change behaviour in the future.  This is definitely not a good situation.  On the other hand, if you mark the attribute with a leading underscore, programmers who are subclassing your code are the ones affected.  They will see the leading underscore and have to figure out whether there's an alternative to using the attribute, what the consequences are to using it, and have to fix their code should that attribute change.  However, since they are writing code that relies intimately on your class anyway understanding the answers to those questions is of benefit to them anyway and something that they are better prepared to do since they are writing code that extends yours rather than simply making use of the services it provides.